### PR TITLE
RFC: add Functors-aware structural gradient

### PR DESCRIPTION
**Pull request by mcabbott**
It's surprisingly simple to add a function which does this:
```julia
julia> model = Chain(Dense(2 => 1, tanh), Dense(1 => 1, bias=false));

julia> withgradient(model, rand(Float32, 2)) do m, x
         sum(abs2, m(x))
       end
(val = 0.035716165f0, grad = ((layers = ((weight = Float32[-0.4241869 -0.16741231], bias = Float32[-0.5529184], σ = nothing), (weight = Float32[-0.04804218;;], bias = nothing, σ = nothing)),), Float32[0.12706584, -0.08858479]))
```
At the moment this is done without touching existing functions, although perhaps `gradient` could be upgraded to do this without breaking anything.

The reason to do so is that this would make Tracker.jl usable again with Flux.jl via Optimisers.jl, and perhaps as a drop-in option for something like https://github.com/FluxML/Flux.jl/pull/2029 . 

The reason not to do this is that Tracker.jl is old and perhaps spreading maintenance effort more thinly is undesirable.

### CODE REVIEW

File: src/back.jl [hunk 1]
```diff
@@ -178,3 +178,60 @@ function jacobian(f, x::AbstractVector)
...
+julia> model = Chain(Dense(2 => 1, tanh), Dense(1 => 1, bias=false));
+
+julia> withgradient(model, rand(Float32, 2)) do m, x
+         sum(abs2, m(x))
+       end
+(val = 0.035716165f0, grad = ((layers = ((weight = Float32[-0.4241869 -0.16741231], bias = Float32[-0.5529184], σ = nothing), (weight = Float32[-0.04804218;;], bias = nothing, σ = nothing)),), Float32[0.12706584, -0.08858479]))
+```
+"""
+function withgradient(f, xs...)
+    pxs = fmap(param, xs; exclude = isnumeric)  # would ideally apply params only to trainable
```
Comment (ToucheSir) [inline src/back.jl:216]: Some variation of `trainable_walk` from https://github.com/FluxML/Optimisers.jl/issues/35#issue-1117418476 could work here.

**Review by ToucheSir — COMMENTED** [review]
I'm pleasantly surprised by how few LOC are needed for this. To make the Optimisers dependency go down better, perhaps it should be thrown under a `@require`? It's unlikely Tracker will drop Requires.jl any time soon, so we might as well make use of that.

File: src/back.jl (hunk 1 above)
Comment (mcabbott) [reply src/back.jl:216]: Thanks, I get lost every time I try to remember how walks work... that looks like the right one.

I guess another option would be not to depend on Optimisers at all, just Functors. Although not tracking non-trainable arrays in Flux probably increases the chances of this just working.

Am not keen on Requires here, seems like a hassle for one tiny package. This already depends on 19 others: https://juliahub.com/ui/Packages/Tracker/cI3wW/0.2.20?page=1

Comment (ToucheSir) [reply src/back.jl:216]: We still need a way to not track non-trainable arrays while still tracking arrays that should be moved on/off GPU in Flux.

I suppose the dep issue isn't a major one in practice, but it may turn a couple of eyebrows. If import times stay mostly the same though, no objections here.

Comment (mcabbott) [reply src/back.jl:216]: Depending on `trainable` does mean raising the floor to Julia 1.6. I think that's fine, anyone on 1.3 has surely accepted freezing all downstream packages by now, we aren't planning to backport bugfixes. 

Zygote of course tries to compute gradients with non-trainable too. But would be nice not to do so.

**Review by ToucheSir — APPROVED** [review]

File: Project.toml [hunk 2]
```diff
@@ -1,33 +1,37 @@
...
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Requires = "ae029012-a4dd-5104-9daa-d747884805df"
 SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
 
 [compat]
-Adapt = "1, 2, 3"
+Adapt = "3"
 DiffRules = "1.4"
+Functors = "0.3.0"
```
Comment (mcabbott) [inline Project.toml:25]: BTW Functors@0.3 required Metalhead#master right now. With that, the example from https://fluxml.ai/Optimisers.jl/dev/#Usage-with-[Flux.jl](https://github.com/FluxML/Flux.jl) runs, and has half the TTFG of Zygote:
```julia
julia> let 
       Random.seed!(1)
       model = Metalhead.ResNet(18) |> gpu  # define a model to train
       image = rand(Float32, 224, 224, 3, 1) |> gpu;  # dummy data
       @show sum(model(image));  # dummy loss function

       rule = Optimisers.Adam()  # use the Adam optimiser with its default settings
       state = Optimisers.setup(rule, model);  # initialise this optimiser's momentum etc.

       @time _, (∇model, _) = Tracker.withgradient(model, image) do m, x  # calculate the gradients
         sum(m(x))
       end;

       state, model = Optimisers.update(state, model, ∇model);
       @show sum(model(image));
       Base.summarysize(∇model)
       end
sum(model(image)) = 1.2527118f0
 19.638126 seconds (39.40 M allocations: 3.444 GiB, 44.46% gc time, 87.70% compilation time)
sum(model(image)) = -4792.643f0
46767520
```
compared to, for Zygote, this:
```
sum(model(image)) = 1.2527118f0
 47.450042 seconds (73.94 M allocations: 5.419 GiB, 36.40% gc time, 93.23% compilation time)
sum(model(image)) = -19.776657f0
46765720
```
But something is wrong, as the final loss differs.

Comment (mcabbott) [reply Project.toml:25]: Looking on the bright side, I guess with this it would be fairly easy to add checks to Flux's tests, comparing what Zygote thinks about each layer to what Tracker thinks. Any which disagree are cause for concern.

Comment (ToucheSir) [reply Project.toml:25]: Yes, and I can already see that being helpful for Metalhead since we see the occasional odd gradient anomaly.

**Comment by mcabbott** [comment]
Is it too weird to leave `gradient` returning tracked arrays and not recursing, unlike `withgradient`?

It would be easy to upgrade it of course, I am a bit scared of breaking things people rely on, half of which I'm quite sure is untested. Might be fairly safe, if it still returns tracked arrays?

And possibly `withgradient` should too, for consistency? Or would that require dozens of second derivative tests?

**Comment by ToucheSir** [comment]
If the worry is about things with similar names not behaving as similarly, perhaps the name could be changed? e.g. steal `explicit_withgradient` from that Flux PR, or (more ugly but unambigious) `withgradient_unwrap`. Customizing unwrapping via a boolean flag could work too. Lastly, gating unwrapping behind a completely distinct interface like https://github.com/JuliaDiff/AbstractDifferentiation.jl/ is an option (and lines up with potential longer-term goals of unifying AD interfaces).

**Comment by mcabbott** [comment]
Thanks for thoughts. Maybe it's OK, using this package is clearly a bit of an adventure... you cannot blindly hope arbitrary 2nd derivatives will work.

File: Project.toml (hunk 2 above)
Comment (mcabbott) [reply Project.toml:25]: Diffractor:
```
sum(model(image)) = 1.2527118f0
 15.313321 seconds (35.07 M allocations: 2.822 GiB, 1.97% gc time, 98.18% compilation time)
sum(model(image)) = -19.776482f0
19384064
```
